File: mock-sse-server.js
/**
 * Mock SSE Server for Testing API Sync
 * Port: 3000
 * Endpoint: /api/crm/stream
 */

const express = require('express');
const app = express();
const PORT = 3000;

// Enable CORS
app.use((req, res, next) => {
    res.header('Access-Control-Allow-Origin', '*');
    res.header('Access-Control-Allow-Headers', '*');
    next();
});

// SSE endpoint - streams CRM data
app.get('/api/crm/stream', (req, res) => {
    console.log(`[${new Date().toISOString()}] New SSE connection from ${req.ip}`);
    
    res.setHeader('Content-Type', 'text/event-stream');
    res.setHeader('Cache-Control', 'no-cache');
    res.setHeader('Connection', 'keep-alive');
    
    // Send connected event
    res.write(`event: connected\n`);
    res.write(`data: {"status": "connected", "message": "Successfully connected to CRM stream"}\n\n`);
    
    // Send initial batch of 10 records
    const initialRecords = [];
    for (let i = 1; i <= 10; i++) {
        initialRecords.push({
            id: i,
            customer_name: `Customer ${i}`,
            email: `customer${i}@example.com`,
            status: i % 2 === 0 ? 'active' : 'pending',
            total_orders: Math.floor(Math.random() * 100),
            total_spent: Math.floor(Math.random() * 10000) / 100,
            created_at: new Date(Date.now() - Math.random() * 30 * 24 * 60 * 60 * 1000).toISOString()
        });
    }
    
    res.write(`event: initial_data\n`);
    res.write(`data: ${JSON.stringify(initialRecords)}\n\n`);
    console.log(`  → Sent initial_data with ${initialRecords.length} records`);
    
    // Send new records every 5 seconds
    let recordId = 11;
    const interval = setInterval(() => {
        if (res.writableEnded) {
            clearInterval(interval);
            return;
        }
        
        const newRecord = {
            id: recordId++,
            customer_name: `Customer ${recordId}`,
            email: `customer${recordId}@example.com`,
            status: recordId % 2 === 0 ? 'active' : 'pending',
            total_orders: Math.floor(Math.random() * 100),
            total_spent: Math.floor(Math.random() * 10000) / 100,
            created_at: new Date().toISOString()
        };
        
        res.write(`event: new_data\n`);
        res.write(`data: ${JSON.stringify(newRecord)}\n\n`);
        console.log(`  → Sent new_data record #${newRecord.id}`);
    }, 5000);
    
    // Cleanup on disconnect
    req.on('close', () => {
        clearInterval(interval);
        console.log(`[${new Date().toISOString()}] SSE connection closed`);
    });
});

// Health check endpoint
app.get('/health', (req, res) => {
    res.json({
        status: 'healthy',
        service: 'Mock SSE Server',
        timestamp: new Date().toISOString()
    });
});

// Start server
app.listen(PORT, () => {
    console.log('='.repeat(80));
    console.log('MOCK SSE SERVER STARTED');
    console.log('='.repeat(80));
    console.log(`Port: ${PORT}`);
    console.log(`SSE Endpoint: http://localhost:${PORT}/api/crm/stream`);
    console.log(`Health Check: http://localhost:${PORT}/health`);
    console.log('='.repeat(80));
    console.log('Waiting for connections...\n');
});

// Graceful shutdown
process.on('SIGINT', () => {
    console.log('\n' + '='.repeat(80));
    console.log('SHUTTING DOWN MOCK SERVER');
    console.log('='.repeat(80));
    process.exit(0);
});

File: mock_sse_server.py
"""
Mock SSE Server for Testing API Sync
Simulates the CRM API at http://localhost:3000/api/crm/stream
"""
from flask import Flask, Response
import json
import time
import random
from datetime import datetime, timedelta

app = Flask(__name__)

# Sample deal data
ACCOUNTS = ["Account A", "Account B", "Account C", "Account D", "Account E"]
STAGES = ["Prospecting", "Needs Analysis", "Proposal", "Negotiation", "Closed Won", "Closed Lost"]
DEAL_TYPES = ["New Business", "Existing Business", "Renewal"]

def generate_deal():
    """Generate a random deal record"""
    deal_id = f"deal-{random.randint(1000, 9999)}"
    return {
        "type": "new_record",
        "record_type": "deal",
        "data": {
            "id": deal_id,
            "deal_name": f"Deal {random.randint(1, 100)} - Q4",
            "amount": random.randint(10000, 150000),
            "stage": random.choice(STAGES),
            "type": random.choice(DEAL_TYPES),
            "closing_date": (datetime.now() + timedelta(days=random.randint(1, 90))).strftime("%Y-%m-%d"),
            "account_name": random.choice(ACCOUNTS),
            "created_time": (datetime.now() - timedelta(days=random.randint(1, 30))).isoformat() + "Z",
            "modified_time": datetime.now().isoformat() + "Z"
        },
        "timestamp": datetime.now().isoformat() + "Z"
    }

@app.route('/api/crm/stream')
def sse_stream():
    """SSE endpoint that streams deal data - supports MULTIPLE concurrent connections"""
    def event_stream():
        client_id = random.randint(1000, 9999)
        print(f"🎉 Client #{client_id} connected to SSE stream!")
        try:
            # Send initial deals
            for i in range(3):
                deal = generate_deal()
                yield f"data: {json.dumps(deal)}\n\n"
                print(f"📤 [Client #{client_id}] Sent deal #{i+1}: {deal['data']['deal_name']}")
                time.sleep(1)
            
            # Continue sending deals every 3 seconds
            counter = 4
            while True:
                time.sleep(3)
                deal = generate_deal()
                yield f"data: {json.dumps(deal)}\n\n"
                print(f"📤 [Client #{client_id}] Sent deal #{counter}: {deal['data']['deal_name']}")
                counter += 1
                
        except GeneratorExit:
            print(f"👋 Client #{client_id} disconnected from SSE stream")
    
    return Response(event_stream(), mimetype='text/event-stream')

@app.route('/api/crm/deals')
def get_deals():
    """REST endpoint that returns a list of deals"""
    deals = [generate_deal()["data"] for _ in range(10)]
    return {"success": True, "data": deals, "count": len(deals)}

@app.route('/')
def index():
    return """
    <h1>Mock CRM API Server</h1>
    <p>Endpoints available:</p>
    <ul>
        <li><a href="/api/crm/stream">/api/crm/stream</a> - SSE streaming endpoint</li>
        <li><a href="/api/crm/deals">/api/crm/deals</a> - REST endpoint (JSON)</li>
    </ul>
    """

if __name__ == '__main__':
    print("="*60)
    print("🚀 Starting Mock CRM API Server")
    print("="*60)
    print("SSE Stream: http://localhost:3000/api/crm/stream")
    print("REST API:   http://localhost:3000/api/crm/deals")
    print("="*60)
    print()
    app.run(host='0.0.0.0', port=3000, debug=False, threaded=True)
